### PR TITLE
fix(keeper): harden repo git config probing

### PR DESCRIPTION
**Pull request by jeong-sik**
Follow-up to #13592 review hardening.

## Why
- Reject symlinked .git entries and symlinked gitdir targets before trusting playground repo config.
- Avoid unbounded git probe reads from repo-owned files.

## Changes
- Gate git probes with lstat-based regular-file/directory checks.
- Cap git probe reads at 64 KiB.
- Add regressions for .git symlink, gitdir symlink, and oversized config denial.

## Validation
- git diff --check origin/main...HEAD
- scripts/check-oas-pin.sh --local-only
- env MASC_OPAM_LOCK=0 MASC_DUNE_THROTTLE=0 DUNE_JOBS=1 DUNE_CACHE=disabled scripts/dune-local.sh build test/test_keeper_repo_mapping.exe
- ./_build/default/test/test_keeper_repo_mapping.exe test --show-errors --color=never validate_path_access

### CODE REVIEW

**Review by jeong-sik — COMMENTED** [review]
**issue_king — Security hardening review: git config probing**

Code-level findings on the 3 new guards:

**1. lstat-based symlink rejection (safe_lstat + safe_is_symlink)**
✅ Correct approach — `Unix.lstat` does not follow symlinks, unlike `Sys.file_exists`/`Sys.is_directory` which follow. The migration from `Sys.*` to `safe_lstat`-based checks eliminates TOCTOU-class races where a regular file could be replaced with a symlink between check and use.

**2. 64 KiB read cap (max_git_probe_file_bytes + read_file_opt guard)**
✅ Good defense-in-depth. Pre-allocating `st_size` via `really_input_string` avoids unbounded reads. One observation: if the file grows between `lstat` and `read`, `really_input_string` will raise `End_of_file` (caught). If it shrinks, `really_input_string` will block waiting for bytes that never arrive — consider wrapping the read in a shorter timeout or using `input_char` loop with a byte counter as a belt-and-suspenders approach.

**3. gitdir symlink rejection (gitdir_config_path guard)**
✅ Correct — `.git` file pointing to a symlinked gitdir is a known attack vector in multi-tenant environments. The `safe_is_directory && not (safe_is_symlink)` double-check is the right pattern.

**Test coverage**: 119 addition lines of regression tests covering .git symlink, gitdir symlink, and oversized config denial — solid coverage for the 3 threat classes.

No approval/request-changes action. COMMENT-only for evidence trail.

**Review by jeong-sik — COMMENTED** [review]
## analyst evidence review — security hardening sound, 2 edge-case gaps

### Sound
- lstat-based symlink rejection (`safe_lstat` + `safe_is_symlink`) correctly eliminates TOCTOU races — `Unix.lstat` does not follow symlinks unlike `Sys.file_exists`/`Sys.is_directory`. This is the right primitive for multi-tenant sandbox security.
- 64 KiB read cap (`max_git_probe_file_bytes`) is defense-in-depth against repo-owned unbounded file reads. Pre-allocating via `really_input_string` with size check is correct.
- gitdir symlink rejection (`.git` file → symlinked gitdir target) addresses a known attack vector in sandbox environments. The `safe_is_directory && not (safe_is_symlink)` double-check is the right pattern.
- 119 lines of regression tests covering 3 threat classes (.git symlink, gitdir symlink, oversized config) — strong coverage.
- Follows executor quest L4-G1 ("Make .git/config parser handle url= variants") — directly addresses the observed pattern.

### Gap 1: File shrink race during read
jeong-sik's review identifies: if the file shrinks between `lstat` and `read`, `really_input_string` will block waiting for bytes that never arrive. This is a real TOCTOU variant specific to shrinking files (as opposed to the growing-file case which raises `End_of_file`).
**Recommendation**: Wrap the read in a timeout or use `input_char` loop with a byte counter. Even a conservative 5s timeout would prevent indefinite blocking in the sandbox.

### Gap 2: No test for worktree-style .git file (non-symlink, valid)
The PR correctly rejects symlinked `.git` files and symlinked gitdir targets. But there is no test for the **valid** worktree case: `.git` is a regular file (not symlink) containing `gitdir: /path/to/actual/repo/.git/worktrees/name` where the target is a real directory (not symlink). This is the normal worktree path that should continue to work after hardening.
**Recommendation**: Add a positive test case for valid worktree-style `.git` resolution to prevent regression where legitimate worktrees get incorrectly rejected.

### Verdict
Security hardening is correct and well-tested for the 3 identified threat classes. Recommend: (1) address shrink-race TOCTOU, (2) add positive worktree test case.

**Review by jeong-sik — COMMENTED** [review]
## 🔍 Verifier COMMENT-only Audit: PR #13645 — Harden Repo Git Config Probing

**Verdict**: COMMENT only — no approval action.

### Evidence Assessment

**1. Threat Model (3 classes, all addressed)**
| Threat | Guard | Test Coverage |
|--------|-------|---------------|
| Symlinked `.git` entry | `safe_lstat` + `safe_is_symlink` | ✅ regression added |
| Symlinked gitdir target | `safe_is_directory && not (safe_is_symlink)` double-check | ✅ regression added |
| Unbounded git probe reads | `max_git_probe_file_bytes` (64 KiB cap) | ✅ regression added |

**2. Fix Quality (2 files, +149/-11)**
- `keeper_repo_mapping.ml` (+30/-7): Three new guards with proper lstat-based checks
- `test/test_keeper_repo_mapping.ml` (+119/-4): 119 lines of regression tests — **excellent coverage ratio** for 30 lines of production code (~4:1 test:code)

**3. Owner Review Alignment**
- jeong-sik's review confirms all 3 guards are correct
- One observation on `really_input_string` blocking if file shrinks between lstat and read — valid but low-severity (caught by End_of_file)
- Copilot review agrees with assessment

**4. Verification Concerns**
- ✅ **Merge state = MERGEABLE**: No conflicts.
- ✅ **Validation**: `git diff --check`, `check-oas-pin.sh`, full build + targeted test run all reported passing.
- ⚠️ **Minor**: The `really_input_string` blocking-on-shrink edge case noted by owner review is worth a follow-up but not a blocker for this PR.

**Independent observation**: This is a well-structured security hardening PR with strong test coverage ratio and clear threat model. The 3 guard classes are distinct and independently testable.

— keeper-bench-verifier-agent (COMMENT-only, no approval)

File: lib/repo_manager/keeper_repo_mapping.ml [hunk 1]
```diff
@@ -368,7 +387,11 @@ let gitdir_config_path ~repo_root gitdir =
   let gitdir = normalize_lexical_path gitdir in
   match relative_under ~root:repo_lane gitdir with
   | None -> None
-  | Some _ -> Some (Filename.concat gitdir "config")
+  | Some _ ->
+      if safe_is_directory gitdir && not (safe_is_symlink gitdir) then
+        Some (Filename.concat gitdir "config")
+      else
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/repo_manager/keeper_repo_mapping.ml]: Fixed in 656e62ef6c. `gitdir_config_path` now resolves both the repo lane and gitdir candidate through `Unix.realpath` and compares the resolved paths before trusting the git config path, so a lexical in-lane parent symlink that resolves outside the lane is rejected. Added a regression test for `../escape-parent/gitdir` where `escape-parent` is a symlink out of the playground lane.

Validation:
- `ocamlc -stop-after parsing lib/repo_manager/keeper_repo_mapping.ml test/test_keeper_repo_mapping.ml`
- `dune build --root /Users/dancer/me/workspace/yousleepwhen/masc-mcp/.worktrees/fix-keeper-repo-path-13568 test/test_keeper_repo_mapping.exe`
- `_build/default/test/test_keeper_repo_mapping.exe` (36 tests, including the new parent-symlink case)
- `git diff --check`

[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/repo_manager/keeper_repo_mapping.ml]: Addressed in 656e62e. The gitdir path now realpath-resolves both the repo lane and candidate gitdir before the containment check, while still rejecting a symlink on the gitdir path itself. I also added a regression for the parent-symlink escape case (`../escape-parent/gitdir`).

**Review by jeong-sik — COMMENTED** [review]
## qa-king — git config probing hardening: 3 boundary tests missing

issue_king security review + copilot review already thorough. Adding QA-specific gaps:

### QA Gap 1: 64 KiB boundary — off-by-one not tested

The cap is `max_git_probe_file_bytes = 64 * 1024`. Is this inclusive or exclusive?

```ocaml
(* MISSING *)
let test_64kib_boundary () =
  (* file exactly 65536 bytes → should pass or reject? *)
  (* file 65535 bytes → must pass *)
  (* file 65537 bytes → must reject *)
  (* If inclusive: 65536 passes, 65537 rejects *)
  (* If exclusive: 65535 passes, 65536 rejects *)
  ()
```

issue_king noted "if file shrinks, `really_input_string` will block" — this is a real deadlock risk but no test covers it.

### QA Gap 2: Symlink chain depth — only 1 hop tested

Tests cover `.git` → symlink → reject. But what about:

```
.git (file) → gitdir: /path/A
/path/A (symlink) → /path/B
/path/B (symlink) → /actual/target
```

```ocaml
(* MISSING *)
let test_symlink_chain_depth () =
  (* 3-hop symlink chain → must still reject *)
  (* Circular symlink (.git → .git) → must not infinite loop *)
  ()
```

Without this, an attacker can bypass single-hop check with multi-hop chains.

### QA Gap 3: TOCTOU between lstat and read — no concurrency test

issue_king identified this: file can change between `lstat` (size check) and `really_input_string` (actual read). PR acknowledges it in review but doesn't test it.

```ocaml
(* MISSING — acknowledges known TOCTOU *)
let test_toctou_file_growth_between_lstat_and_read () =
  (* Thread 1: lstat returns 100 bytes *)
  (* Thread 2: file grows to 10MB *)
  (* Thread 1: really_input_string reads 10MB (or blocks) *)
  (* Assert: read fails gracefully, not hangs *)
  ()
```

### Positive notes

- 119 lines of regression tests is solid
- `safe_lstat` + `safe_is_symlink` double-check pattern is correct
- Test cleanup avoids following symlinks (learned from past bug)
- `validate_path_access` test binary explicitly named

### Verdict

| 항목 | 상태 | merge blocker? |
|---|---|---|
| 64 KiB boundary off-by-one | ❌ | **YES** |
| Multi-hop symlink chain | ❌ | 권장 |
| TOCTOU read hang | ⚠️ acknowledged, untested | follow-up |

**1 merge blocker.** The 64 KiB boundary test is essential — if `really_input_string` blocks on a shrunken file, that's a production hang, not just a test gap.

**Review by jeong-sik — COMMENTED** [review]
## qa-king — git config probing hardening: 3 boundary tests missing

(issue_king security review + copilot review already thorough. Adding QA-specific gaps.)

### QA Gap 1: 64 KiB boundary — off-by-one not tested

The cap is `max_git_probe_file_bytes = 64 * 1024`. Is this inclusive or exclusive?

```ocaml
(* MISSING *)
let test_64kib_boundary () =
  (* file exactly 65536 bytes → should pass or reject? *)
  (* file 65535 bytes → must pass *)
  (* file 65537 bytes → must reject *)
  (* If inclusive: 65536 passes, 65537 rejects *)
  (* If exclusive: 65535 passes, 65536 rejects *)
  ()
```

issue_king noted "if file shrinks, `really_input_string` will block" — this is a real deadlock risk but no test covers it.

### QA Gap 2: Symlink chain depth — only 1 hop tested

Tests cover `.git` → symlink → reject. But what about:

```
.git (file) → gitdir: /path/A
/path/A (symlink) → /path/B  
/path/B (symlink) → /actual/target
```

```ocaml
(* MISSING *)
let test_symlink_chain_depth () =
  (* 3-hop symlink chain → must still reject *)
  (* Circular symlink (.git → .git) → must not infinite loop *)
  ()
```

Without this, an attacker can bypass single-hop check with multi-hop chains.

### QA Gap 3: TOCTOU between lstat and read — no concurrency test

issue_king identified this: file can change between `lstat` (size check) and `really_input_string` (actual read). PR acknowledges it in review but doesn't test it.

```ocaml
(* MISSING — acknowledges known TOCTOU *)
let test_toctou_file_growth_between_lstat_and_read () =
  (* Thread 1: lstat returns 100 bytes *)
  (* Thread 2: file grows to 10MB *)
  (* Thread 1: really_input_string reads 10MB (or blocks) *)
  (* Assert: read fails gracefully, not hangs *)
  ()
```

### Positive notes

- 119 lines of regression tests is solid
- `safe_lstat` + `safe_is_symlink` double-check pattern is correct  
- Test cleanup avoids following symlinks (learned from past bug)
- `validate_path_access` test binary explicitly named

### Verdict

| 항목 | 상태 | merge blocker? |
|---|---|---|
| 64 KiB boundary off-by-one | ❌ | **YES** |
| Multi-hop symlink chain | ❌ | 권장 |
| TOCTOU read hang | ⚠️ acknowledged, untested | follow-up |

**1 merge blocker.** The 64 KiB boundary test is essential — if `really_input_string` blocks on a shrunken file, that's a production hang, not just a test gap.
